File: clients/java/src/main/java/com/acteon/client/models/Attachment.java
package com.acteon.client.models;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * An attachment with explicit metadata and base64-encoded data.
 */
public class Attachment {
    private String id;
    private String name;
    private String filename;
    @JsonProperty("content_type")
    private String contentType;
    @JsonProperty("data_base64")
    private String dataBase64;

    /** No-arg constructor for Jackson deserialization. */
    public Attachment() {}

    /**
     * Creates an attachment.
     *
     * @param id          user-set identifier for referencing in chains
     * @param name        human-readable display name
     * @param filename    filename with extension
     * @param contentType MIME type (e.g., "application/pdf")
     * @param dataBase64  base64-encoded file content
     */
    public Attachment(String id, String name, String filename, String contentType, String dataBase64) {
        this.id = id;
        this.name = name;
        this.filename = filename;
        this.contentType = contentType;
        this.dataBase64 = dataBase64;
    }

    public String getId() { return id; }
    public void setId(String id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public String getFilename() { return filename; }
    public void setFilename(String filename) { this.filename = filename; }

    public String getContentType() { return contentType; }
    public void setContentType(String contentType) { this.contentType = contentType; }

    public String getDataBase64() { return dataBase64; }
    public void setDataBase64(String dataBase64) { this.dataBase64 = dataBase64; }
}

File: crates/core/src/attachment.rs
use serde::{Deserialize, Serialize};

/// An attachment to be included with an action dispatch.
///
/// Each attachment carries a user-defined `id` that makes it referenceable
/// across actions and sub-chains, a human-readable `name`, the original
/// `filename` with extension, a MIME `content_type`, and the file content
/// as `base64`-encoded data.
///
/// Providers that support attachments (email, Slack, Discord, webhook) resolve
/// these at dispatch time; providers that don't simply ignore them.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct Attachment {
    /// User-set identifier for referencing this attachment in chains.
    pub id: String,
    /// Human-readable display name.
    pub name: String,
    /// Filename with extension (e.g. `"report.pdf"`).
    pub filename: String,
    /// MIME content type (e.g. `"application/pdf"`).
    pub content_type: String,
    /// `Base64`-encoded file content.
    pub data_base64: String,
}

/// A fully resolved attachment with decoded binary data.
///
/// Built by the gateway after decoding `base64` content. Passed to providers
/// via [`DispatchContext`](crate::attachment::ResolvedAttachment).
#[derive(Debug, Clone)]
pub struct ResolvedAttachment {
    /// User-set identifier.
    pub id: String,
    /// Human-readable display name.
    pub name: String,
    /// Filename with extension.
    pub filename: String,
    /// MIME content type.
    pub content_type: String,
    /// Decoded binary content.
    pub data: Vec<u8>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn attachment_serde_roundtrip() {
        let attachment = Attachment {
            id: "att-1".into(),
            name: "Hello File".into(),
            filename: "hello.txt".into(),
            content_type: "text/plain".into(),
            data_base64: "SGVsbG8gV29ybGQ=".into(),
        };
        let json = serde_json::to_string(&attachment).unwrap();
        assert!(json.contains("\"id\":\"att-1\""));
        assert!(json.contains("\"name\":\"Hello File\""));
        let back: Attachment = serde_json::from_str(&json).unwrap();
        assert_eq!(back.id, "att-1");
        assert_eq!(back.name, "Hello File");
        assert_eq!(back.filename, "hello.txt");
        assert_eq!(back.content_type, "text/plain");
        assert_eq!(back.data_base64, "SGVsbG8gV29ybGQ=");
    }

    #[test]
    fn empty_attachments_vec_deserializes_from_missing_field() {
        // Simulates backward compatibility: old payloads without "attachments"
        let json = r#"[]"#;
        let attachments: Vec<Attachment> = serde_json::from_str(json).unwrap();
        assert!(attachments.is_empty());
    }
}
